_examples/01_one_account_one_region/README.md

```diff
@@ -0,0 +1,155 @@
+│   |   └── terraform.tfvars        <- component configuration
+│   └── vpc
+│       └── terraform.tfvars
+└── terraform.tfvars                <- root configuration
+```
+
+The `live/` directory is the root of our Terragrunt configuration hierarchy.
+It contains a single `terraform.tfvars` file, which we will call
+the **root configuration**. The root configuration is optional, but allows 
+us to define general configuration that applies to all components in our infrastructure
+in a single place.
+
+The root contains a child directory for each environment in the infrastructure
+(e.g. production, staging), which we will call **environment directories**.
+The directory structure shown above contains two environment directories:
+`production` and `staging`.
+Environment directories do not normally contain any files. Instead, they are
+the location where Terragrunt commands such as `plan-all` and `apply-all`
+are executed to instantiate all of the infrastructure components for the
+given environment.
+
+Environment directories contain a child directory for each component of 
+the infrastructure, which we will call **component directories**.
+For example, a component might be a single app, a single MySQL DB, or a single VPC.
+The directory structure shown above contains 4 component directories: `frontend-app` and
+`vpc` in each environment.
+
+Each component directory contains a single `terraform.tfvars` file, which
+defines the configuration for that specific component. We will call this
+`terraform.tfvars` file the **component configuration**. The directory
+structure shown above contains 4 component configurations: one for
+each component in each environment.
+In a component configuration, we can use the `include` stanza to instruct 
+Terragrunt to locate and copy configuration from the root configuration file. 
+This allows us to keep our configurations DRY. Review the example
+code in this repo to see how this works in practice.
+
+## Each infrastructure component is defined in a Terraform module
+In this example, our infrastructure contains two components: a VPC and a frontend_app.
+Each of these components is defined in its own Terraform module:
+[conorgil/terragrunt_sample_vpc](https://github.com/conorgil/terragrunt_sample_vpc)
+and
+[conorgil/terragrunt_sample_frontend_app](https://github.com/conorgil/terragrunt_sample_frontend_app)
+respectively.
+
+There are a few important things to note about the Terraform modules:
+1. The code in each of these repos is 100% regular plain old Terraform code. These
+   Terraform modules can be used, without modification, in any project that does
+   not use Terragrunt and instead uses Terarform directly.
+1. Each module defines a variable for any value that might change between invocations
+   of the module. For example, values that might change between production and staging
+   environments.
+1. Each module enables a backend for storing remote state. The backend intentionally

```

Comment:
@dpetzel This is a great question. The short answer is yes. As I have written the example, the remote state is defined inside the Terraform module. I decided to write the example as I did because of [this discussion](https://github.com/gruntwork-io/terragrunt/issues/169#issuecomment-299199548) in #169.

The long answer is that I think this is an open question that is being discussed as to what the "best practice" (see #230). 

As I was writing the example, it did feel "wrong" to hardcode some things into the Terraform modules; specifically, the `backend` and `terraform_remote_state`. I think that it felt "wrong" because I want to apply the dependency injection pattern to Terraform modules and hardcoding a `backend` and `terraform_remote_state` violates that pattern.

I'll ramble about `backend` and leave my thoughts on `terraform_remote_state` for another issue/time. As you said, if you define a backend in a module, then that module cannot be used by another part of your team with a different backend; they would have to change the source code to change the backend. When using plain old Terraform, the backend is defined once in the `main.tf` file. One or more Terraform modules are called using the classic syntax and passing in any values they might need as variables. The Terraform modules do not define any backend because it was defined in the `main.tf` file already and there is only a single state file. With Terragrunt, we want each component to maintain its own state file, which means that each component is defined in its own directory and also needs to configure its own backend. As you point out, this could be done in a `backend.tf` file file along side the component configuration instead of in the Terraform module itself. This would certainly make the Terraform module more generic and follow the dependency injection pattern more closely since other teams could simply define their own `backend.tf` file and the module would work without change. There is some discussion in #230 about challenges supporting this feature natively in Terragrunt, but if you were to just follow this pattern manually, then I think it should work fine. There would be more files to maintain in the Terragrunt configuration repo, but it could be worth it if generic reusable Terraform modules are important for your team. If you try this (or already do this) let me know how it works out!

I think the counter argument is one that you actually highlighted in [this comment](https://github.com/gruntwork-io/terragrunt/issues/169#issuecomment-300942218).
> your not building generic multi-use type of modules, but rather modules that are implementing a very specific configuration, unique to your scenario.

If the module is unique to each scenario, then reusability is not the main goal and the additional overhead of maintaining `backend.tf` files is likely not worth it if all groups in the team use the same backend type anyways.